tests/libwnbd_tests/test_disk_actions.cpp

```diff
@@ -457,3 +458,164 @@ TEST(TestUnmap, TestHardUnmapFallback) {
+        sizeof(sptd),
+        &bytesReturned,
+        NULL);
+
+    EXPECT_NE(Result, 0) << "Error sending command: " << GetLastError();
+    ASSERT_EQ(sptd.ScsiStatus, 0) << "Command returned with error status: " << std::hex << sptd.ScsiStatus;
+    CloseHandle(hDevice);
+
+    WNBD_IO_REQUEST ExpWnbdRequest = { 0 };
+    ExpWnbdRequest.RequestType = WnbdReqTypePersistResIn;
+    ExpWnbdRequest.Cmd.PersistResIn.ServiceAction = RESERVATION_ACTION_READ_RESERVATIONS;
+    ASSERT_TRUE(WnbdDaemon.ReqLog.HasEntry(ExpWnbdRequest));
+
+    WNBD_USR_STATS UserspaceStats = { 0 };
+    WnbdGetUserspaceStats(WnbdDisk, &UserspaceStats);
+    ASSERT_EQ(UserspaceStats.PersistResInErrors, 0);
+}
+
+TEST(TestPersistentReservations, TestPersistentReserveOut) {
+    auto InstanceName = GetNewInstanceName();
+
+    MockWnbdDaemon WnbdDaemon(
+        InstanceName,
+        DefaultBlockCount,
+        DefaultBlockSize,
+        false,
+        false,
+        false,
+        false,
+        true
+    );
+
+    WnbdDaemon.Start();
+    WNBD_CONNECTION_INFO ConnectionInfo = { 0 };
+    ASSERT_FALSE(WnbdShow(InstanceName.c_str(), &ConnectionInfo))
+        << "couldn't retrieve WNBD disk info";
+
+    std::string DiskPath = GetDiskPath(InstanceName);
+    PWNBD_DISK WnbdDisk = WnbdDaemon.GetDisk();
+
+    UCHAR cdb[16];
+    UCHAR data[24];
+    UCHAR sense[32];
+
+    ZeroMemory(cdb, sizeof(cdb));
+    ZeroMemory(data, sizeof(data));
+    ZeroMemory(sense, sizeof(sense));
+
+    cdb[0] = SCSIOP_PERSISTENT_RESERVE_OUT;
+
+    HANDLE hDevice = CreateFileA(
+        DiskPath.c_str(),
+        GENERIC_READ | GENERIC_WRITE,
+        FILE_SHARE_READ | FILE_SHARE_WRITE,
+        NULL,
+        OPEN_EXISTING,
+        0,
+        NULL);
+
+    ASSERT_NE(hDevice, INVALID_HANDLE_VALUE)
+        << "Error opening device: " << GetLastError();
+
+    SCSI_PASS_THROUGH_DIRECT sptd;
+    ZeroMemory(&sptd, sizeof(sptd));
+    sptd.Length = sizeof(sptd);
+    sptd.CdbLength = 6;
+    sptd.DataIn = SCSI_IOCTL_DATA_IN;
+    sptd.DataBuffer = data;
+    sptd.DataTransferLength = sizeof(data);
+    sptd.TimeOutValue = 2;
+    sptd.Cdb[0] = SCSIOP_PERSISTENT_RESERVE_OUT;
+    sptd.Cdb[1] = RESERVATION_ACTION_CLEAR;
+    sptd.Cdb[2] = RESERVATION_TYPE_EXCLUSIVE;
+    DWORD bytesReturned = 0;
+    BOOL Result = DeviceIoControl(
+        hDevice,
+        IOCTL_SCSI_PASS_THROUGH_DIRECT,
+        &sptd,
+        sizeof(sptd),
+        &sptd,
+        sizeof(sptd),
+        &bytesReturned,
+        NULL);
+
+    EXPECT_NE(Result, 0) << "Error sending command: " << GetLastError();
+    ASSERT_EQ(sptd.ScsiStatus, 0) << "Command returned with error status: " << std::hex << sptd.ScsiStatus;
+    CloseHandle(hDevice);
+
+    //// TODO: find out why ReqLog.HasEntry returns false for the request we send
+    /*WNBD_IO_REQUEST ExpWnbdRequest = { 0 };
+    ExpWnbdRequest.RequestType = WnbdReqTypePersistResOut;
+    ExpWnbdRequest.Cmd.PersistResOut.ServiceAction = RESERVATION_ACTION_CLEAR;
```

Comment:
This check fails because we aren't filling in all the expected command fields, we'll need this:

ExpWnbdRequest.Cmd.PersistResOut.Type = RESERVATION_TYPE_EXCLUSIVE;